interp/interp.go

```diff
@@ -361,30 +366,64 @@ func (interp *Interpreter) resizeFrame() {
 	interp.frame.data = data
 }
 
-func (interp *Interpreter) main() *node {
-	interp.mutex.RLock()
-	defer interp.mutex.RUnlock()
-	if m, ok := interp.scopes[mainID]; ok && m.sym[mainID] != nil {
-		return m.sym[mainID].node
-	}
-	return nil
-}
-
 // Eval evaluates Go code represented as a string. Eval returns the last result
 // computed by the interpreter, and a non nil error in case of failure.
 func (interp *Interpreter) Eval(src string) (res reflect.Value, err error) {
 	return interp.eval(src, "", true)
 }
 
-// EvalPath evaluates Go code located at path. EvalPath returns the last result
-// computed by the interpreter, and a non nil error in case of failure.
-func (interp *Interpreter) EvalPath(path string) (res reflect.Value, err error) {
-	// TODO(marc): implement eval of a directory, package and tests.
-	b, err := ioutil.ReadFile(path)
-	if err != nil {
-		return res, err
+// EvalPath evaluates Go code located at path, running tests specified by tfe
+// which is the test filter regular expression. If tfe is empty, the main function
+// is evaluated. EvalPath returns the last result computed by the interpreter,
+// and a non nil error in case of failure.
+func (interp *Interpreter) EvalPath(path string, skiptest bool) (res reflect.Value, err error) {
+	if isFile(path) {
+		b, err := ioutil.ReadFile(path)
+		if err != nil {
+			return res, err
+		}
+		return interp.eval(string(b), path, false)
+	}
+	_, err = interp.importSrc("main", path, skiptest)
+	return res, err
+}
+
+// Symbols returns a map of interpreter exported symbol values for the given path.
+func (interp *Interpreter) Symbols(path string) map[string]reflect.Value {
+	m := map[string]reflect.Value{}
+
+	interp.mutex.RLock()
+	if interp.scopes[path] == nil {
+		interp.mutex.RUnlock()
+		return m
+	}
+	sym := interp.scopes[path].sym
+	interp.mutex.RUnlock()
+
+	for n, s := range sym {
+		if !canExport(n) {
+			// Skip private non-exported symbols.
+			continue
+		}
+		switch s.kind {
+		case constSym:
+			m[n] = s.rval
+		case funcSym:
+			m[n] = genFunctionWrapper(s.node)(interp.frame)
+		case varSym:
+			if i := s.index; i >= 0 {
+				m[n] = interp.frame.data[i]
+			} else {
+				m[n] = genValue(s.node)(interp.frame)
+			}
```

Comment:
as discussed by chat, this might not be needed

Reply:
ok